app/wasm/bindings/msg.go

```diff
@@ -4,28 +4,47 @@ import sdk "github.com/cosmos/cosmos-sdk/types"
 
 type OsmosisMsg struct {
 	/// Contracts can create denoms, namespaced under the contract's address.
-	//A contract may create any number of independent sub-denoms.
+	/// A contract may create any number of independent sub-denoms.
 	CreateDenom *CreateDenom `json:"create_denom,omitempty"`
-	/// Contracts can mint native tokens for an existing denom
-	/// namespaced under the contract's address.
+	/// Contracts can change the admin of a denom that they are the admin of.
+	ChangeAdmin *ChangeAdmin `json:"change_admin,omitempty"`
+	/// Contracts can mint native tokens for an existing factory denom
+	/// that they are the admin of.
 	MintTokens *MintTokens `json:"mint_tokens,omitempty"`
+	/// Contracts can burn native tokens for an existing factory denom
+	/// that they are the admin of.
+	/// Currently, the burn from address must be the admin contract.
+	BurnTokens *BurnTokens `json:"burn_tokens,omitempty"`
 	/// Swap over one or more pools
 	Swap *SwapMsg `json:"swap,omitempty"`
 }
 
+/// CreateDenom creates a new factory denom, of denomination:
+/// factory/{creating contract address}/{Subdenom}
+/// Subdenom can be of length at most 44 characters, in [0-9a-zA-Z./]
+/// The (creating contract address, subdenom) pair must be unique.
+/// The created denom's admin is the creating contract address,
+/// but this admin can be changed using the UpdateAdmin binding.
 type CreateDenom struct {
-	/// Sub_denoms (nonces) are validated as part of the full denomination.
-	/// Can be up to 128 - prefix length (currently 7) - bech32 address length (4 (osmo) + 39) - number of separators (2) =
-	/// 76 "alphanumeric" (https://github.com/cosmos/cosmos-sdk/blob/2646b474c7beb0c93d4fafd395ef345f41afc251/types/coin.go#L677)
-	/// characters long.
-	/// Empty sub-denoms are valid. The token will then be prefix + contract address, i.e. "factory/<bech32 address>/"
-	SubDenom string `json:"sub_denom"`
+	Subdenom string `json:"subdenom"`
+}
+
+type ChangeAdmin struct {
+	Denom           string `json:"denom"`
+	NewAdminAddress string `json:"new_admin_address"`
 }
 
 type MintTokens struct {
-	SubDenom  string  `json:"sub_denom"`
-	Amount    sdk.Int `json:"amount"`
-	Recipient string  `json:"recipient"`
+	Denom         string  `json:"denom"`
+	Amount        sdk.Int `json:"amount"`
+	MintToAddress string  `json:"mint_to_address"`
+}
+
+type BurnTokens struct {
+	Denom  string  `json:"denom"`
+	Amount sdk.Int `json:"amount"`
+	// BurnFromAddress must be set to "" for now.
+	BurnFromAddress string `json:"burn_from_address"`
 }
```

Comment:
Need to decide if Rust side we include BurnFromAddress and require it being empty on golang side or keep it empty for now.

The go side can handle either way. Not sure if bindings in Rust use traits vs macros vs direct structs. For traits / macros we can remove it and add it in optionally later. For a direct struct, depends on if fields are private and its generated via constructor, or if fields are public and can be directly put in.

Reply:
I think we should add it, and then throw on the Go side